Add a JWT testing guide. (#744)
OWASP/wstg: document/4-Web_Application_Security_Testing/06-Session_Management_Testing/10-Testing_JSON_Web_Tokens.md

## Test Objectives

- Determine whether the JWTs expose sensitive information.
- Determine whether the JWTs can be tampered with or modified.

## How to Test

### Overview

JWTs are are made up of three components:

- The header
- The payload (or body)
- The signature

Each component is Base64 encoded, and they are separated by periods (`.`). Note that the Base64 encoding used in a JWT strips out the equals signs (`=`), so you may need to add these back in to decode the sections.

### Analyse the Contents

#### Header

The header of a defines the type of the token (typically `JWT`), and the algorithm used for the signature. A example decoded header is shown below:

```json
{
  "alg": "HS256",
  "typ": "JWT"
}
```

There are three main types of algorithms that are used to calculate the signatures:

| Algorithm | Description |
|-----------|-------------|
| HSxxx | HMAC using a secret key and SHA-xxx. |
| RSxxx and PSxxx | Public key signature using RSA. |
| ESxxx | Public key signature using ECDSA. |

> There are also a wide range of [other algorithms](https://www.iana.org/assignments/jose/jose.xhtml#web-signature-encryption-algorithms) which may be used for encrypted tokens (JWEs), although these are less common.s

Reviewer: @rbsec I think there is an extra 's' at the end of paragraph.

Author: Fixed

The validity of the JWT should also be reviewed, based on the `iat`, `nbf` and `exp` claims, to determine that:

- The JWT has a reasonable lifespan for the application.
- Expired tokens are rejected by the application.

### Signature Verification

One of the most serious vulnerabilities encountered with JWTs is when the application fails to validate that the signature is correct. This usually occurs when a developer uses a function such as the NodeJS `jwt.deode()` function, which simply decodes the body of the JWT, rather than `jwt.verify()`, which verifies the signature before decoding the JWT.

This can be easily tested for by modifying the body of the JWT without changing anything in the header or signature, submitting it in a request to see if the application accepts it.

#### The None Algorithm

As well as the public key and HMAC-based algorithms, the JWT specification also defines a signature algorithm called `none`. As the name suggests, this means that there is no signature for the JWT, allowing it to be modified.

> This can be tested by modifying the signature algorithm (`alg`) in the JWT header to `none`, and resubmitting the JWT.

Reviewer: @rbsec @kingthorin Is it good to add an example?
also mention that signature must remove and the last '.' must remains.

Author: Good spot, I'd forgotten that. Fixed in 42454ca and added an example.

# Testing JSON Web Tokens

|ID          |
|------------|
|WSTG-SESS-010|

## Summary

JSON Web Tokens (JWTs) are cryptographically signed JSON tokens, intended to share claims between systems. They are are frequently used as authentication or session tokens, particularly on REST APIs.

JWTs are a common source of vulnerabilities, both in how they are in implemented in applications, and in the underlying libraries. As they are used for authentication, a vulnerability can easily result in a complete compromise of the application.

## Test Objectives

- Determine whether the JWTs expose sensitive information.
- Determine whether the JWTs can be tampered with or modified.

## How to Test

### Overview

JWTs are are made up of three components:

- The header
- The payload (or body)
- The signature

Each component is Base64 encoded, and they are separated by periods (`.`). Note that the Base64 encoding used in a JWT strips out the equals signs (`=`), so you may need to add these back in to decode the sections.

### Analyse the Contents

#### Header

> The header of a defines the type of the token (typically `JWT`), and the algorithm used for the signature. A example decoded header is shown below:

Reviewer 2: ```suggestion
The header of a token defines the type of that token (typically `JWT`), and the algorithm used for the signature. An example decoded header is shown below:
```

#### Payload

The payload of the JWT contains the actual data. An example payload is shown below:

```json
{
  "username": "admininistrator",
  "is_admin": true,
  "iat": 1516239022,
  "exp": 1516242622
}
```

The payload is it not usually encrypted, so review it to determine whether there is any sensitive of potentially inappropriate data included within it.

This JWT includes the username and administrative status of the user, as well as two standard claims (`iat` and `exp`). These claims are defined in [RFC 5719](https://tools.ietf.org/html/rfc7519#section-4.1), a brief summary of them is given in the table below:

| Claim | Full Name | Description |
|-------|-----------|-------------|
| `iss` | Issuer | The identity of the party who issued the token. |
| `iat` | Issued At | The Unix timestamp of when the token was issued. |
| `nbf` | Not Before | The Unix timestamp of earliest date that the token can be used. |
| `exp` | Expires | The Unix timestamp of when the token expires. |

#### Signature

> The signature is calculated using the algorithm defined in the JWT header, and then Base64 encoded and appened to the token. Modifying any part of the JWT should cause the signature to be invalid, and the token to be rejected by the server.

Reviewer 2: ```suggestion
The signature is calculated using the algorithm defined in the JWT header, and then Base64 encoded and appended to the token. Modifying any part of the JWT must cause the signature to be invalid, and the token to be rejected by the server.
```

> The signature is calculated using the algorithm defined in the JWT header, and then Base64 encoded and appened to the token. Modifying any part of the JWT should cause the signature to be invalid, and the token to be rejected by the server.

Reviewer 2: I deleted this by mistake.
```suggestion
The signature is calculated using the algorithm defined in the JWT header, and then Base64 encoded and appended to the token. Modifying any part of the JWT must cause the signature to be invalid, and the token to be rejected by the server.
```

Author: I used "should" here because it's making an assumption that the validation library is doing its job properly - which is often not the case.

Reviewer 2: If the signature was to be checked and comes out as invalid, then it must cause it to ... Otherwise it's flat out ignored. I get what you're saying, just not the case for this paragraph as this is more on how it functions properly.

### Review Usage

As well as being cryptographically secure itself, the JWT also needs to be stored and sent in a secure manner. This should include checks that:

- It is always [sent over encrypted (HTTPS) connections](../09-Testing_for_Weak_Cryptography/03-Testing_for_Sensitive_Information_Sent_via_Unencrypted_Channels.md).
- If it is stored in a cookie, then it should be [marked with appropriate attributes](../06-Session_Management_Testing/02-Testing_for_Cookies_Attributes.md).

> - If it is stored in the browser storage, it should [use the sessionStorage rather than localStorage](../11-Client-side_Testing/12-Testing_Browser_Storage.md).

Reviewer 2: I have thoughts on this, but I feel like it deserves its own issue. A lot of identity providers just use local storage when not using cookies, or basically a framework type of storage (which is something _similar_ to local storage). I'll open an issue on this because it affects more sections of the guide.

Author: This came up a bit in the issue for this PR - but I don't think I really want to got into much detail here about it. In fact, it might be best to remove this recommendation entirely given some of the issues around sessionStorage..

Reviewer 2: Feel free to remove that, no hard feelings on this.

The validity of the JWT should also be reviewed, based on the `iat`, `nbf` and `exp` claims, to determine that:

> - The JWT has a reasonable lifespan for the application.

Reviewer 2: @kingthorin should we link to the ASVS  here, or should we just keep this vague like this.

This can be tested by modifying the signature algorithm (`alg`) in the JWT header to `NoNe`, as shown in the example below:

```json
{
        "alg": "NoNe",
        "typ": "JWT"
}
```

The header and payload are then re-encoded with Base64, and the signature is removed (leaving the trailing period). Using the header above, and the payload listed in the [payload](#payload) section, this would give the following JWT:

```txt
eyJhbGciOiAiTm9OZSIsICJ0eXAiOiAiSldUIn0.eyJ1c2VybmFtZSI6ImFkbWluaW5pc3RyYXRvciIsImlzX2FkbWluIjp0cnVlLCJpYXQiOjE1MTYyMzkwMjIsImV4cCI6MTUxNjI0MjYyMn0.
```

### Weak HMAC Keys

If the JWT is signed using a HMAC-based algorithm (such as HS256), the security of the signature is entirely reliant on the strength of the secret key used in the HMAC.

If the application is using off-the-shelf or open source software, the first step should be go investigate the code, and see whether there is default HMAC signing key that is used.

If there isn't a default, then it may be possible to crack guess or brute-force they key. The simplest way to do this is to use the [crackjwt.py](https://github.com/Sjord/jwtcrack) script, which simply requires the JWT and a dictionary file.

> A more powerful option is to convert convert the JWT into a format that can be used by [John the Ripper](https://github.com/openwall/john) using the [jwt2john.py](https://github.com/Sjord/jwtcrack/blob/master/jwt2john.py) script. John can then be used to carry out much more advanced attacks against the key.

Reviewer 2: ```suggestion
A more powerful option is to convert the JWT into a format that can be used by [John the Ripper](https://github.com/openwall/john) using the [jwt2john.py](https://github.com/Sjord/jwtcrack/blob/master/jwt2john.py) script. John can then be used to carry out much more advanced attacks against the key.
```

> If the JWT is large, it may exceed the maximum size supported by John. This can be worked around by increasing the value of the `SALT_LIMBS` variable in `/src/hmacSHA256_fmt_plug.c` (or the equivalent file for other HMAC formats) and recompiling John.

Reviewer 2: Is there a link for this guidance so that we are not held as an authority on this?

Author: I've added a link to the [GitHub issue](https://github.com/openwall/john/issues/1904)

If this key can be obtained, then it is possible to create and sign arbitrary JWTs, which usually results in a complete compromise of the application.

### HMAC vs Public Key Confusion

If the application uses JWTs with public key based signatures, but does not check that the algorithm is correct, this can potentially exploit this in a signature type confusion attack. In order for this to be successful, the following conditions need to be met:

- The application must use public key (i.e, `RSxxx` or `ESxxx`) to verify the JWT signature.
- The application must not check which algorithm is actually used.
- The public key used to verify the JWT must be available to the attacker.

> The main way that an attacker would be able to obtain the public key is if the application re-uses the same key for both signing JWTs and as part of the TLS certificate. This key can be downloaded from the server using a command such as the following:

Reviewer 2: This line/sentence feels a bit weird.
What do feel about the following suggestions:
1. Number the bullet points above.
2. Mention that the weird or troublesome point is point #2, as the public key must be public, unless they're misusing it (different thing .. haha). From the start of this sentence it feels like the issue is with the public key.
3. Start off by first discussing jwks and then the TLS cert re-use.

Author: I've made some changes to this section in 202a12b to try and clear it up.

```sh
openssl s_client -connect example.org:443 | openssl x509 -pubkey -noout
```

Alternatively, the key may be available from a public file on the site at a common location such as `/.well-known/jwks.json`.

> In order to test this, modify the contents of the JWT, and then use the previously obtained public key to sign the JWT using the `HS256` algorithm. This is often difficult to perform when testing without access to the source code or implementation deatils, because the format of the key must be identical to the one used by the server, so issues such as emptyspace or CRLF encoding may result in the keys not matching.

Reviewer 2: not sure if the empty space was intended, but it was spot on.
```suggestion
In order to test this, modify the contents of the JWT, and then use the previously obtained public key to sign the JWT using the `HS256` algorithm. This is often difficult to perform when testing without access to the source code or implementation details, because the format of the key must be identical to the one used by the server, so issues such as empty space or CRLF encoding may result in the keys not matching.
```

### Attacker Provided Public Key

> The JSON Web Signature (JWS) standard allows the key used to sign the token to be embedded in the header. If the library used to validate the token supports this, and doesn't check the key against a list of approved keys, this allows an attacker to sign an JWT with an arbitrary key that they provide.

Reviewer 2: I felt like this should have been discussed earlier, and since it was mentioned I feel like it needs to. JWTs are almost all JWSs, and I feel like it might create some confusion at this stage for unaware readers. What do you think?

Author: I wonder if mentioning JWS at all here is just opening a can of worms? The whole JWE/JWS/JWT issue is somewhat confusing, and in reality almost everyone just refers to JWTs and ignores the rest of it.

I've slightly amended the wording in 202a12b and linked to the JWS RFC - do you think it needs more than that? If people want to really understand how all the different bits of the JWT spec fit together, they probably need to go and do some of their own reading.

Reviewer 2: Maybe having a tiny note somewhere, just so we don't get the "Why are you talking about JWTs and they are JWSs". We can link to the JOSE protocols as well in the references and call it a day as well. Let's definitely not dive anywhere into it.

Author: I've updated the text to:

> The [JSON Web Signature (JWS) standard](https://tools.ietf.org/html/rfc7515) (which defines the header and signatures used by JWTs) allows the key used to sign the token to be embedded in the header.

And added a link to that RFC in the references. Do you think  it needs anything more than that?

There are a variety of scripts that can be used to do this, such as [jwk-node-jose.py](https://github.com/zi0Black/POC-CVE-2018-0114) or [jwt_tool](https://github.com/ticarpi/jwt_tool).

## Related Test Cases

- [Testing for Sensitive Information Sent via Unencrypted Channels](../09-Testing_for_Weak_Cryptography/03-Testing_for_Sensitive_Information_Sent_via_Unencrypted_Channels.md).
- [Testing for Cookie Attributes](../06-Session_Management_Testing/02-Testing_for_Cookies_Attributes.md).
- [Testing Browser Storage](../11-Client-side_Testing/12-Testing_Browser_Storage.md).

## Remediation

> - Use a secure and up to date library to handle JWTs.

Reviewer 2: We can add something to the lines of "be sure to validate the signature"

> # Testing JSON Web Tokens (JWTs)

Reviewer 3: ```suggestion
# Testing JSON Web Tokens
```

We should do this clarification in the first paragraph, then linking is less convoluted.

|ID          |
|------------|
|WSTG-SESS-010|

## Summary

> - JWTs are cryptographically signed JSON tokens

Reviewer 3: ```suggestion
- JSON Web Tokens (JWTs) are cryptographically signed JSON tokens
```

- Commonly used to authenticate on APIs
- Lots of bad implementations and opportunities for things to go wrong

## Test Objectives

- Determine whether the JWTs expose sensitive information.
- Determine whether the JWTs can be tampered with or modified.

## How to Test

### Overview

- What a JWT is
    - Three b64 encoded parters (header, body, signature)
- Common usage scenarios

### Analyse the Contents

- JWTs are not usually encrypted
- May expose sensitive information

### Review Usage

- How is JWT sent?
- How is JWT stored?
- Does JWT expire?
- Can JWT be invalidated?

### Signature Verification

- Signature may not be verified at all (`jwt.decode()` vs `jwt.verify()`)

> - `none` or `NoNe` algorithms may be allowed

Reviewer 3: RSA56? (That's top of head, I may have the wrong detail, but there is another weak 'standard')

Author: There's actually a whole ton of algorithms that you very rarely see, but most of them are related to JWE (encrypted tokens) rather than the more common JWTs.

[RFC 7518](https://www.rfc-editor.org/rfc/rfc7518.txt) has a list, and there's a slightly clearer table on the [IANA site](https://www.iana.org/assignments/jose/jose.xhtml#web-signature-encryption-algorithms).

I've never seen another other than RS256/HS256/ES256 used, and a full discussion of all the algorithms is probably going to get a bit complicated. Maybe something like "If the JWT is using anything other than RSxxx, HSxxx, or ESxxx then go and look into the algorithm properly" ?

Reviewer 3: Ya a simple statement like that probably covers it

### Weak HMAC Keys

- HMAC uses a secret key
- Convert JWT into crackable format
- Try and crack key
    - jwt2john
    - May need custom compiled john for long JWT (`SALT_LIMBS`)
    - jwtcrack

### HMAC vs RSA Confusion

- Obtain public key
    - May be in `/.well-known/jwks.json`
    - Can be obtained from TLS cert (shouldn't be re-used, but may be)
- Use public key as a secret to calculate a HMAC

### Attacker Provided RSA Key

- JSON Web Signature (JWS) allows embedding of public key in the header
- Sign with our own private key and see if server trusts it

## Related Test Cases

TODO

## Remediation

TODO

## Tools

- [John the Ripper](https://github.com/openwall/john)
- [jwt2john](https://github.com/Sjord/jwtcrack)
- [jwt-cracker](https://github.com/brendan-rius/c-jwt-cracker)

> - [JSON Web Tokens Burp Extension](https://portswigger.net/bappstore/f923cbf91698420890354c1d8958fee6)

Reviewer 3: ```suggestion
- [JSON Web Tokens Burp Extension](https://portswigger.net/bappstore/f923cbf91698420890354c1d8958fee6)
- [ZAP JWT Add-on](https://github.com/SasanLabs/owasp-zap-jwt-addon)
```
